### PR TITLE
[SC-156] Fix `nil` errors and raise more explicit exceptions

### PR DESCRIPTION
**Pull request by linux2647**
## Problem

Two problems:

* There are cases when a Fedex tracking event lacks an address.  Previous code tried to access sub-items in these `nil` address blocks, causing `undefined method `[]' for nil:NilClass` errors
* Errors returned from the Fedex web services API raise generic `Fedex::RateError` exceptions

## Solution

* Use an alternate method to lookup data in an address block that accounts for an address to potentially be `nil`
* Create two categories of exceptions based on the response from the Fedex web services API:
  * **Fault** exceptions that are raised when something like an internal server error response is returned from the API
  * **Error** exceptions that are raised when the request to the API was invalid and needs to be corrected before resubmission
    * **InvalidTrackingNumber** exception, as a subclass of **Error** that explicitly states that a submitted tracking number is invalid

Note: the above exceptions all inherit from the existing `Fedex::RateError` exception to preserve backwards-compatibility.

## Future work

The created exceptions are by no means exhaustive, as there are many different kinds of error codes that can be returned from the Fedex API.  Some error codes have the exact same error description so it is not clear whether these error codes are functionally identical or not.

### CODE REVIEW

File: lib/fedex.rb [hunk 1]
```diff
@@ -52,4 +52,11 @@ module Fedex
   require 'fedex/version'
   #Exceptions: Fedex::RateError
   class RateError < StandardError; end
+
+  # Indicates a general fault in the Fedex web service
+  class TrackingInformationFault < RateError; end
```
Comment (mandemeskel) [inline lib/fedex.rb:57]: Can we call this `FedexServerError` and the other error `FedexRequestError`?

`TrackingInformationFault` is too vague, if I saw this and `TrackingInformationError` I would not know the difference without having to dig up this file and look at the comments. I don't think we should depend on people's knowledge of the nuanced difference between `fault` and `error`. The comments are opaque as well, should just use what you wrote in the PR - also, it's a code smell if you have to use a comment to explain what the error means.

File: lib/fedex.rb [hunk 2]
```diff
@@ -52,4 +52,11 @@ module Fedex
...
+  # Indicates a general error returned from the Fedex web service
+  class TrackingInformationError < RateError; end
+  # A submitted tracking number is invalid
+  class InvalidTrackingNumberError < TrackingInformationError; end
```
Comment (mandemeskel) [inline lib/fedex.rb:61]: love it, thank you 💟

File: lib/fedex/request/tracking_information.rb [hunk 3]
```diff
@@ -36,12 +36,23 @@ def process_request
             Fedex::TrackingInformation.new(details)
           end
         else
-          error_message = if response[:track_reply]
-            response[:track_reply][:notifications][:message]
+          if response[:track_reply]
+            error_message = response[:track_reply][:notifications][:message]
+            # These error codes come from the Fedex Webservices Developer Guide (2019)
+            # (https://www.fedex.com/us/developer/downloads/pdf/2019/FedEx_WebServices_DevelopersGuide_v2019.pdf
+            #  via https://www.fedex.com/en-us/developer/web-services/process.html#documentation).
```
Comment (mandemeskel) [inline lib/fedex/request/tracking_information.rb:43]: nice comment

File: lib/fedex/request/tracking_information.rb [hunk 4]
```diff
@@ -36,12 +36,23 @@ def process_request
...
+            error_code = response[:track_reply][:notifications][:code]
+            if invalid_tracking_number? error_code
+              raise InvalidTrackingNumberError, "[#{error_code}] #{error_message} (#{@package_id})"
```
Comment (mandemeskel) [inline lib/fedex/request/tracking_information.rb:46]: after millennia of suffering and countless lost alerts we have arrived...transparent errors, what a time to be alive!

File: lib/fedex/tracking_information/event.rb [hunk 5]
```diff
@@ -8,11 +8,16 @@ def initialize(details = {})
...
+        # Address may not exist, so we use #dig as it will extract a deeply-nested key or return nil if any intermediate
+        # step is nil. There are cases when :address is not nil but city or state/province code is nil (like a "shipping
+        # details sent" event) so there is precedent for some of these fields to be nil.  See the WSDL provided by
+        # Fedex for more information: https://www.fedex.com/us/developer/downloads/wsdl/2019/standard/TrackService.zip
+        # (via: https://www.fedex.com/en-us/developer/web-services/process.html#documentation)
+        @city                  = details.dig(:address, :city)
+        @state                 = details.dig(:address, :state_or_province_code)
+        @postal_code           = details.dig(:address, :postal_code)
+        @country               = details.dig(:address, :country_code)
+        @residential           = details.dig(:address, :residential) == "true"
```
Comment (mandemeskel) [inline lib/fedex/tracking_information/event.rb:20]: nice!

**Comment by linux2647** [comment]
> can you add tests

So per our BJN conversation, tests don't run.  I've been working on getting them to run, but there is still some work to be done.  I noticed there _are_ tests in FashionThing, so I [added some there](https://github.com/stitchfix/fashionthing/pull/5425), and they are applicable to FashionThing, so it's not out of place for those tests to exist there.  When tests start working again in this repo, I'll add them here.

File: lib/fedex.rb (hunk 1 above)
Comment (linux2647) [reply lib/fedex.rb]: Done

**Comment by linux2647** [comment]
I have pulled in the working rspec code from `master` into this branch and have incorporated the tests from FashionThing into this repo as well.  This means that this branch has tests for the changes made in this branch 😄

File: lib/fedex.rb [hunk 6]
```diff
@@ -52,4 +52,11 @@ module Fedex
   require 'fedex/version'
   #Exceptions: Fedex::RateError
   class RateError < StandardError; end
+
+  # Indicates a general fault in the Fedex web service
+  class FedexInternalServerError < RateError; end
+  # Indicates a general error returned from the Fedex web service
+  class FedexRequestError < RateError; end
```
Comment (mandemeskel) [inline lib/fedex.rb:59]: thanks for making the name changes!

File: spec/lib/fedex/track_spec.rb [hunk 7]
```diff
@@ -47,6 +47,19 @@ module Fedex
         tracking_info.status.should == "Shipment cancelled by sender"
       end
 
+      # The following tests require constructed responses
+      unless real_credentials?
+        it "does not raise an exception when an event's address is missing" do
+          # In the past, if an address block was missing in one of the Fedex tracking events, a
+          # "NoMethodError: undefined method `[]' for nil:NilClass" exception would be raised.
+          tracking_info = fedex.track(options).first
+          tracking_info.should_not be nil
```
Comment (mandemeskel) [inline spec/lib/fedex/track_spec.rb:56]: looks good!

**Pull request merged**